File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/AttestationClientAuth.kt
package com.hopae.eudi.wallet

import com.hopae.eudi.wallet.sdjwt.SecureAreaJwsSigner
import com.hopae.eudi.wallet.spi.KeyHandle
import com.hopae.eudi.wallet.spi.KeyInfo
import com.hopae.eudi.wallet.spi.KeySpec
import com.hopae.eudi.wallet.spi.Rng
import com.hopae.eudi.wallet.spi.SecureArea
import com.hopae.eudi.wallet.spi.SigningAlgorithm
import com.hopae.eudi.wallet.spi.StorageDriver
import com.hopae.eudi.wallet.spi.WalletAttestationProvider
import com.hopae.eudi.wallet.vci.ClientAttestationPop
import com.hopae.eudi.wallet.vci.ClientAuthProvider
import com.hopae.eudi.wallet.vci.WalletClientAuth
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * HAIP attestation-based client authentication, wired from [WalletPorts.walletAttestation]. Holds one
 * persistent **wallet instance key** (created in the [secureArea] and remembered via [storage]; bound into
 * every WUA's `cnf`) and, on each request, produces the `OAuth-Client-Attestation[-PoP]` headers.
 *
 * HAIP §4.4.1 (unlinkability): a WUA is fetched **fresh per authorization server** and **never reused across
 * issuers** — [wuaByAudience] caches one WUA per `aud` so a single issuer's PAR + token reuse it, but a
 * different issuer always triggers a fresh fetch. The `sub` is the (non-instance-unique) [clientId].
 *
 * Residual (documented, not fixed here): because the instance key is persistent, the WUA `cnf` is the same
 * across issuers, so colluding issuers could still correlate via the key. Full unlinkability needs a per-use
 * key batch (HAIP wallet attestation batch) — a later refinement.
 */
internal class AttestationClientAuth(
    override val clientId: String,
    private val provider: WalletAttestationProvider,
    private val secureArea: SecureArea,
    private val storage: StorageDriver,
    private val rng: Rng,
    private val clock: () -> Long,
) : ClientAuthProvider {

    private val keyMutex = Mutex()
    private val wuaMutex = Mutex()
    private var instance: KeyInfo? = null
    private val wuaByAudience = mutableMapOf<String, String>()

    override suspend fun headers(audience: String): List<Pair<String, String>> {
        val key = instanceKey()
        val wua = walletAttestation(audience, key)
        val pop = ClientAttestationPop(SecureAreaJwsSigner(secureArea, key.handle, key.algorithm), clientId, rng, clock)
        return WalletClientAuth(clientId, wua, pop).headers(audience)
    }

    /** The persistent wallet instance key: created once, its alias remembered so it survives restarts. */
    private suspend fun instanceKey(): KeyInfo = keyMutex.withLock {
        instance?.let { return it }
        val stored = storage.get(COLLECTION, INSTANCE_KEY)
        val info = if (stored != null) {
            val handle = KeyHandle(secureArea.id, stored.decodeToString())
            KeyInfo(handle, SigningAlgorithm.ES256, secureArea.publicKey(handle))
        } else {
            secureArea.createKey(KeySpec(secureArea = secureArea.id)).also {
                storage.put(COLLECTION, INSTANCE_KEY, it.handle.alias.encodeToByteArray())
            }
        }
        instance = info
        info
    }

    private suspend fun walletAttestation(audience: String, key: KeyInfo): String = wuaMutex.withLock {
        wuaByAudience[audience] ?: provider.walletAttestation(key).also { wuaByAudience[audience] = it }
    }

    private companion object {
        const val COLLECTION = "wallet-provider"
        const val INSTANCE_KEY = "instance-key"
    }
}

File: swift/Sources/Wallet/AttestationClientAuth.swift
import OpenID4VCI
import SdJwt
import WalletAPI

/// HAIP attestation-based client authentication, wired from `WalletPorts.walletAttestation`. Holds one
/// persistent wallet instance key (created in the `secureArea` and remembered via `storage`; bound into
/// every WUA's `cnf`) and, on each request, produces the `OAuth-Client-Attestation[-PoP]` headers.
///
/// HAIP §4.4.1 (unlinkability): a WUA is fetched fresh per authorization server and never reused across
/// issuers — `wuaByAudience` caches one WUA per `aud` so a single issuer's PAR + token reuse it, but a
/// different issuer always triggers a fresh fetch. The `sub` is the (non-instance-unique) `clientId`.
///
/// Residual (documented): the instance key is persistent, so the WUA `cnf` is the same across issuers —
/// colluding issuers could still correlate. Full unlinkability needs a per-use key batch (a later refinement).
final class AttestationClientAuth: ClientAuthProvider {
    let clientId: String
    private let provider: any WalletAttestationProvider
    private let secureArea: any SecureArea
    private let storage: any StorageDriver
    private let rng: any Rng
    private let clock: () -> Int64

    private var instance: KeyInfo?
    private var wuaByAudience: [String: String] = [:]

    init(clientId: String, provider: any WalletAttestationProvider, secureArea: any SecureArea,
         storage: any StorageDriver, rng: any Rng, clock: @escaping () -> Int64) {
        self.clientId = clientId
        self.provider = provider
        self.secureArea = secureArea
        self.storage = storage
        self.rng = rng
        self.clock = clock
    }

    func headers(audience: String) async throws -> [(String, String)] {
        let key = try await instanceKey()
        let wua = try await walletAttestation(audience: audience, key: key)
        let pop = ClientAttestationPop(
            signer: SecureAreaJwsSigner(area: secureArea, key: key.handle, algorithm: key.algorithm),
            clientId: clientId, rng: rng, now: clock)
        return try await WalletClientAuth(clientId: clientId, attestationJwt: wua, pop: pop).headers(audience: audience)
    }

    /// The persistent wallet instance key: created once, its alias remembered so it survives restarts.
    private func instanceKey() async throws -> KeyInfo {
        if let instance { return instance }
        let info: KeyInfo
        if let stored = try await storage.get(collection: Self.storeCollection, key: Self.storeKey) {
            let handle = KeyHandle(secureArea: secureArea.id, alias: String(decoding: stored, as: UTF8.self))
            info = KeyInfo(handle: handle, algorithm: .es256, publicKey: try await secureArea.publicKey(key: handle))
        } else {
            let created = try await secureArea.createKey(spec: KeySpec(secureArea: secureArea.id))
            try await storage.put(collection: Self.storeCollection, key: Self.storeKey, value: [UInt8](created.handle.alias.utf8))
            info = created
        }
        instance = info
        return info
    }

    private func walletAttestation(audience: String, key: KeyInfo) async throws -> String {
        if let cached = wuaByAudience[audience] { return cached }
        let wua = try await provider.walletAttestation(keyInfo: key)
        wuaByAudience[audience] = wua
        return wua
    }

    private static let storeCollection = "wallet-provider"
    private static let storeKey = "instance-key"
}
